### ISSUE TITLE
Negative decimals when forceByteArrayEncoding is set to true

### ISSUE THREAD

**Issue by macalbert** [elastacloud/parquet-dotnet#275]
**Version:** Parquet.Net v2.0.1

I'm using parquet.net to storage data on S3, and after I'm using [redshift spectrum](https://aws.amazon.com/redshift/spectrum/?nc1=h_ls) to make queries over this data.

To using decimals on redshift spectrum, I need to set **forceByteArrayEncoding** to **true**, otherwise an error like the following will be raise.

```
[Amazon](500310) Invalid operation: S3 Query Exception (Fetch)
Details: 

 -----------------------------------------------
  error:  S3 Query Exception (Fetch)
  code:      15001
  context:   Task failed due to an internal error. File 'https://s3-eu-west-1.amazonaws.com/.../data.parquet  has an incompatible Parquet schema for column 's3://...
  query:     5322247
  location:  dory_util.cpp:686
  process:   query0_41 [pid=25133]
```

But when I'm using a decimal where **forByteArrayEncoding** is set to **true**, the negative values like **-1** are read like **1,56**.

Here a test that details this behaivor:

```csharp
using Microsoft.VisualStudio.TestTools.UnitTesting;
using System.Linq;

namespace Tuitele.UnitTest
{ 
    [TestClass]
    public class TestClass 
    {
        [TestMethod]
        public void Write_loads_of_decimals_all_negatives()
        {
            const int _numberOfItemsToGenerate = 100;
            const decimal _negativeValue = -1;

            Parquet.Data.DataSet ds1 = new Parquet.Data.DataSet(new Parquet.Data.DataField<int>("id"),
                                                                new Parquet.Data.DecimalDataField("valueA", 10, 2, forceByteArrayEncoding: false),
                                                                new Parquet.Data.DecimalDataField("valueB", 10, 2, forceByteArrayEncoding: true));
            for (int n = 0; n < _numberOfItemsToGenerate; n++)
            {
                ds1.Add(n, _negativeValue, _negativeValue);
            }

            Parquet.Data.DataSet ds2 = Parquet.ParquetReader.Read(this.WriteParquet(ds1));

            for (int i = 0; i < ds2.Count; i++)
            {
                Parquet.Data.Row row = ds2.ElementAt(i);

                Assert.IsTrue(decimal.Compare(row.Get<decimal>(1), _negativeValue) == 0, $"got {row.Get<decimal>(1)} at position {i}"); // valueA forceByteArrayEncoding false
                Assert.IsTrue(decimal.Compare(row.Get<decimal>(2), _negativeValue) == 0, $"got {row.Get<decimal>(2)} at position {i}"); // valueB forceByteArrayEncoding true
            }
        }

        private System.IO.Stream WriteParquet(Parquet.Data.DataSet dataSet)
        {
            System.IO.Stream stream = new System.IO.MemoryStream();

            using (var writer = new Parquet.ParquetWriter(stream))
            {
                writer.Write(dataSet);
            }

            stream.Position = 0;

            return stream;
        }
    }
}
```

**Comment by aloneguid** [comment]
I can confirm Apache Spark reads it as 1.56 as well, seems to be serious.

**Comment by aloneguid** [comment]
I sort of got to the bottom of it. The problem is that Java and C# uses different implementations of byte array conversion to/from `BigInteger` which is used internally for decimal encoding:

https://docs.oracle.com/javase/7/docs/api/java/math/BigInteger.html#toByteArray()

https://msdn.microsoft.com/en-us/library/system.numerics.biginteger.tobytearray(v=vs.110).aspx

The different is mostly endianness, and we handle it at the moment by reversing byte array when writing to Parquet, however sign bit is somehow not trasferrable between C# and Java, trying to understand why.

**Comment by macalbert** [comment]
Maybe this can help? [https://stackoverflow.com/a/36701951/541877](https://stackoverflow.com/a/36701951/541877)

**Comment by aloneguid** [comment]
Thanks, but that's how we do it at the moment, so positive values are encoded properly. The problem is that BigInteger is encoded into a fixed byte array, so let's say we're encoding `-1.2` with precision 10 and scale 2:

- the output buffer has to be 5 bytes
- the value encodes to 1 byte [136]

therefore the output is [0, 0, 0, 0, 136]. But, for some reason parquet-mr doesn't think it is. There must be a sign bit missing somewhere, but i can't really understand where.

**Comment by aloneguid** [comment]
@xhochy I wonder if you've got a nice hint for us ;)

**Comment by macalbert** [comment]
And this ? [https://msdn.microsoft.com/en-us/library/system.numerics.biginteger.aspx](https://msdn.microsoft.com/en-us/library/system.numerics.biginteger.aspx#Remarks)

> **Working with Byte Arrays and Hexadecimal Strings**
[...]
The BigInteger structure assumes that negative values are stored by using two's complement representation. Because the BigInteger structure represents a numeric value with no fixed length, the BigInteger(Byte[]) constructor always interprets the most significant bit of the last byte in the array as a sign bit. To prevent the BigInteger(Byte[]) constructor from confusing the two's complement representation of a negative value with the sign and magnitude representation of a positive value, positive values in which the most significant bit of the last byte in the byte array would ordinarily be set should include an additional byte whose value is 0. For example, 0xC0 0xBD 0xF0 0xFF is the little-endian hexadecimal representation of either -1,000,000 or 4,293,967,296. Because the most significant bit of the last byte in this array is on, the value of the byte array would be interpreted by the BigInteger(Byte[]) constructor as -1,000,000. To instantiate a BigInteger whose value is positive, a byte array whose elements are 0xC0 0xBD 0xF0 0xFF 0x00 must be passed to the constructor. The following example illustrates this.

**Comment by macalbert** [comment]
Perhaps this class can help to reverse the bytes: [BitConverter Class](https://msdn.microsoft.com/en-us/library/system.bitconverter(v=vs.110).aspx)

`If systems sending and receiving data can have different endianness, always transmit data in a particular order. This means that the order of bytes in the array may have to be reversed either before sending them or after receiving them. A common convention is to transmit data in network byte order (big-endian order). The following example provides an implementation for sending an integer value in network byte order`

```csharp
using System;

public class Example
{
   public static void Main()
   {
      int value = 12345678;
      byte[] bytes = BitConverter.GetBytes(value);
      Console.WriteLine(BitConverter.ToString(bytes));

      if (BitConverter.IsLittleEndian)
         Array.Reverse(bytes); 

      Console.WriteLine(BitConverter.ToString(bytes));
      // Call method to send byte stream across machine boundaries.

      // Receive byte stream from beyond machine boundaries.
      Console.WriteLine(BitConverter.ToString(bytes));
      if (BitConverter.IsLittleEndian)
         Array.Reverse(bytes);

      Console.WriteLine(BitConverter.ToString(bytes));
      int result = BitConverter.ToInt32(bytes, 0);
      Console.WriteLine("Original value: {0}", value);
      Console.WriteLine("Returned value: {0}", result);
   }
}
// The example displays the following output on a little-endian system:
//       4E-61-BC-00
//       00-BC-61-4E
//       00-BC-61-4E
//       4E-61-BC-00
//       Original value: 12345678
//       Returned value: 12345678
```

**Comment by aloneguid** [comment]
@macalbert unfortunately that's not the case - it's a question what to put in extra bytes. When you set precision in bigdecimal the buffer size for number is fixed, the question is what to put in bytes which are not used, that's where reader gets confused.
Parquet.Net reads negative decimals with no problem, just confirmed.

**Comment by aloneguid** [comment]
When I write value `-1` with Spark in legacy mode with precision 38 and scale 18 i get the following byte buffer back:

`[255, 255, 255, 255, 255, 255, 255, 255, 242, 31, 73, 76, 88, 156, 0, 0]` (need to be reversed for C#'s BigInteger). Not sure how this comes out.

**Comment by aloneguid** [comment]
Managed to force Spark to write precision 10 and scale 2, for value -1 we get `[255, 255, 255, 255, 156]` which is a bit odd. I think byte 255 is written to keep the negative bit on `1000 0000` but why `156` is there for -1?

**Comment by aloneguid** [comment]
Got it, -1 is written as -100 (scale is 2). Therefore in binary it's `0001 0100 (28)`. Adding negative bit we get `1001 0100` which is 156.

**Comment by aloneguid** [comment]
All good now:

```csharp
var ds = new DataSet(new DecimalDataField("dec", 10, 2, true));
ds.Add(-1.0m);
ds.Add(-1.23m);
ds.Add(-23233.12m);
ds.Add(-999999.999m);
```

is read by Spark as:

```
+----------+
|       dec|
+----------+
|     -1.00|
|     -1.23|
| -23233.12|
|-999999.99|
+----------+
```

**Comment by aloneguid** [comment]
Released as hotfix in v2.1.1, please reopen if it still doesn't work ;)

**Comment by macalbert** [comment]
Thanks @aloneguid!! Tested in redshift spectrum and reads it as -1!
